--- a/arch/x86/Makefile
+++ b/arch/x86/Makefile
@@ -176,7 +176,7 @@
 
 ifdef CONFIG_X86_NATIVE_CPU
-        KBUILD_CFLAGS += -march=native
+        KBUILD_CFLAGS += -march=native -mtune=native
         KBUILD_RUSTFLAGS += -Ctarget-cpu=native
 endif
--- a/arch/x86/Kconfig.cpu
+++ b/arch/x86/Kconfig.cpu
@@ -247,7 +247,7 @@
 
 config CC_HAS_MARCH_NATIVE
 	# This flag might not be available in cross-compilers:
-	def_bool $(cc-option, -march=native)
+	def_bool $(cc-option, -march=native -mtune=native)
